// @itexpert-dev/base-universal-table/test/karma-webpack.config.js
module.exports = [{
    entry: './src/main.ts',
    output: {
        filename: 'main.js',
        path: '/serve',
        publicPath: '/'
    },
    devtool: 'source-map',
    resolve:{
      extensions: ['.tsx', '.ts', '.js']
    },
    module: {
        rules: [
            {
                test: /\.ts$/,
                loader: 'awesome-typescript-loader',
                exclude: '/mode_modules/'
            }
            ,{
                test: /\.pug$/,
                use: `pug-html-loader?basedir=${__dirname}`
            },{
                test: /\.css$/,
                use: [
                    'css-to-string-loader',
                    'style-loader',
                    'css-loader'
                ]
            },{
                test: /\.styl$/,
                use: [
                    'css-to-string-loader',
                    'style-loader',
                    'css-loader',
                    'stylus-loader'
                ]
            },{
                test: /\.html$/,
                use: 'html-loader'
            },{
                test   : /\.(ttf|eot|svg|woff(2)?)(\?[a-z0-9=&.]+)?$/,
                use : 'file-loader'
            },{
                test: /\.svg$/,
                use: 'file-loader'
            }
        ]
    },
    devServer: {
        inline: true,
        historyApiFallback: true
    }
}]
