// @itexpert-dev/base-universal-table/dist/directives/common/baseUniversalTableDirective.js
"use strict";
var BaseUniversalTableDirective = (function () {
    function BaseUniversalTableDirective(componentResolver, viewContainer, tableCtrl) {
        this.componentResolver = componentResolver;
        this.viewContainer = viewContainer;
        this.tableCtrl = tableCtrl;
    }
    BaseUniversalTableDirective.prototype.resolveTablePart = function (tablePart) {
        var tableCtrl = this.tableCtrl;
        //as any because tableCtrl.TableComponents not dictionary
        var component = tableCtrl.tableComponents[tablePart];
        var componentFactory = tableCtrl.config.componentsStorage.getComponent(component);
        return this.componentResolver.resolveComponentFactory(componentFactory);
    };
    BaseUniversalTableDirective.prototype.initTablePart = function (tablePart) {
        var resolvedComponent = this.resolveTablePart(tablePart);
        var componentRef = this.viewContainer.createComponent(resolvedComponent);
        return componentRef;
    };
    BaseUniversalTableDirective.prototype.resolveCellByType = function (type) {
        var tableCtrl = this.tableCtrl;
        var componentName = tableCtrl.config.cellTypeToComponentMap.get(type);
        var componentFactory = tableCtrl.config.cellsStorage.getComponent(componentName);
        return this.componentResolver.resolveComponentFactory(componentFactory);
    };
    return BaseUniversalTableDirective;
}());
exports.BaseUniversalTableDirective = BaseUniversalTableDirective;
