// @itexpert-dev/base-universal-table/dist/services/universalTable/baseUniversalTableConfigurator/baseUniversalTableConfigurator.js
"use strict";
var components_key_value_storage_1 = require("@itexpert-dev/components-key-value-storage");
var key_value_storage_1 = require("@itexpert-dev/key-value-storage");
var UniversalTableConfigurator = (function () {
    function UniversalTableConfigurator() {
        this.componentsStorage = new components_key_value_storage_1.ComponentsKeyValueStorage();
        this.cellsStorage = new components_key_value_storage_1.ComponentsKeyValueStorage();
        this.cellTypeToComponentMap = new key_value_storage_1.KeyValueStorage();
        this.cellTypeToComponentMap.add({ key: 'primary', value: 'primaryCell' });
    }
    UniversalTableConfigurator.prototype.setDefaultInitTableState = function (newState) {
        this._defaultInitTableState = newState;
    };
    UniversalTableConfigurator.prototype.getDefaultInitTableState = function () {
        return this._defaultInitTableState;
    };
    UniversalTableConfigurator.prototype.setDefaultComponents = function (components) {
        this._defaultComponents = components;
    };
    UniversalTableConfigurator.prototype.getDefaultComponents = function () {
        return this._defaultComponents;
    };
    UniversalTableConfigurator.prototype.setDefaultLimitSteps = function () {
        var steps = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            steps[_i] = arguments[_i];
        }
        this._defaultLimitSteps = steps;
    };
    UniversalTableConfigurator.prototype.getDefaultLimitSteps = function () {
        return this._defaultLimitSteps;
    };
    return UniversalTableConfigurator;
}());
exports.UniversalTableConfigurator = UniversalTableConfigurator;
