// @itexpert-dev/base-universal-table/src/directives/universalTableHeaders/universalTableHeaders.directive.ts
import {
    Directive,
    ComponentFactoryResolver,
    ViewContainerRef,
    Input,
    Injector
} from "@angular/core";
import {UniversalTableCtrl} from "../../services/universalTable/universalTableCtrl";
import {baseUniversalTableHeadersDirectiveConfig} from "./baseUniversalTableHeadersDirectiveConfig.const";
import {BaseUniversalTableHeadersDirective} from "./BaseUniversalTableHeadersDirective";
import {UniversalTableHeaderData} from './UniversalTableHeaderData'

@Directive(baseUniversalTableHeadersDirectiveConfig)
class UniversalTableHeadersDirective extends BaseUniversalTableHeadersDirective{
    @Input() headerIndex: number;
    @Input() headerData: any;

    constructor(
        componentResolver: ComponentFactoryResolver,
        viewContainer: ViewContainerRef,
        tableCtrl: UniversalTableCtrl,
        injector: Injector
    ){
        super(
            tableCtrl,
            componentResolver,
            viewContainer,
            injector
        );
    }
}

export {UniversalTableHeadersDirective, BaseUniversalTableHeadersDirective, baseUniversalTableHeadersDirectiveConfig, UniversalTableHeaderData}
