// karma-typescript/integration-tests@latest/tmp/node/core/core-module-tester.js
"use strict";
var assert = require("assert");
var buffer = require("buffer");
var console = require("console");
var constants = require("constants");
var crypto = require("crypto");
var domain = require("domain");
var events = require("events");
var http = require("http");
var os = require("os");
var util = require("util");
var CoreModuleTester = (function () {
    function CoreModuleTester() {
    }
    CoreModuleTester.prototype.testAssert = function () {
        return assert.equal;
    };
    CoreModuleTester.prototype.testBuffer = function () {
        return new buffer.Buffer("string");
    };
    CoreModuleTester.prototype.testConsole = function () {
        return console.log;
    };
    CoreModuleTester.prototype.testConstants = function () {
        return constants.ENOENT;
    };
    CoreModuleTester.prototype.testCrypto = function () {
        return crypto.createHash('md5').update("string").digest('hex');
    };
    CoreModuleTester.prototype.testDomain = function () {
        return domain.create();
    };
    CoreModuleTester.prototype.testEvents = function () {
        return new events.EventEmitter();
    };
    CoreModuleTester.prototype.testHttp = function () {
        return http.Agent;
    };
    CoreModuleTester.prototype.testOs = function () {
        return os.platform();
    };
    CoreModuleTester.prototype.testUtil = function () {
        return util.isNumber(42);
    };
    return CoreModuleTester;
}());
exports.CoreModuleTester = CoreModuleTester;
