// @itexpert-dev/components-key-value-storage/dist/componentsKeyValueStorage/componentsKeyValueStorage.js
"use strict";
var key_value_storage_1 = require("@itexpert-dev/key-value-storage");
var ComponentsKeyValueStorage = (function () {
    function ComponentsKeyValueStorage() {
        this.storage = new key_value_storage_1.KeyValueStorage();
    }
    ComponentsKeyValueStorage.prototype.addComponent = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        for (var _a = 0, params_1 = params; _a < params_1.length; _a++) {
            var param = params_1[_a];
            var fixedParam = {
                value: param.component,
                force: param.force,
                key: param.componentName
            };
            this.storage.add(fixedParam);
        }
    };
    ComponentsKeyValueStorage.prototype.removeComponent = function (key) {
        this.storage.remove(key);
    };
    ComponentsKeyValueStorage.prototype.getComponent = function (key) {
        return this.storage.get(key);
    };
    return ComponentsKeyValueStorage;
}());
exports.ComponentsKeyValueStorage = ComponentsKeyValueStorage;
