// @itexpert-dev/i-crud-repository/src/main.ts
import {IConnectorApi} from './ICrudRepository/ICrudRepositoryApi'

import {ICrudRepositoryCreateRequest} from './ICrudRepository/create/request/ICrudRepositoryCreateRequest'
import {ICrudRepositoryCreateResponse} from './ICrudRepository/create/response/ICrudRepositoryCreateResponse'

import {
    ICrudRepositoryReadRequest,
    ICrudRepositoryReadRequestFilterBlockFormat,
    CrudRepositoryReadRequestFilterBlockOperator,
    ICrudRepositoryReadRequestFieldFormat
} from './ICrudRepository/read/request/ICrudRepositoryReadRequest'
import {ICrudRepositoryReadResponse} from './ICrudRepository/read/response/ICrudRepositoryReadResponse'

import {ICrudRepositoryUpdateRequest} from './ICrudRepository/update/request/ICrudRepositoryUpdateRequest'
import {ICrudRepositoryUpdateResponse} from './ICrudRepository/update/response/ICrudRepositoryUpdateResponse'

import {ICrudRepositoryDeleteRequest} from './ICrudRepository/delete/request/ICrudRepositoryDeleteRequest'
import {ICrudRepositoryDeleteResponse} from './ICrudRepository/delete/response/ICrudRepositoryDeleteResponse'

import {ICrudRepositoryMetadataRequest} from './ICrudRepository/metadata/request/ICrudRepositoryMetadataRequest'
import {ICrudRepositoryMetadataResponse, ICrudRepositoryMetadataResponseProperty} from './ICrudRepository/metadata/response/ICrudRepositoryMetadataResponse'

export {
    IConnectorApi,
    ICrudRepositoryCreateRequest,
    ICrudRepositoryCreateResponse,

    ICrudRepositoryReadRequest,
    ICrudRepositoryReadRequestFilterBlockFormat,
    CrudRepositoryReadRequestFilterBlockOperator,
    ICrudRepositoryReadRequestFieldFormat,

    ICrudRepositoryReadResponse,
    ICrudRepositoryUpdateRequest,
    ICrudRepositoryUpdateResponse,
    ICrudRepositoryDeleteRequest,
    ICrudRepositoryDeleteResponse,

    ICrudRepositoryMetadataRequest,
    ICrudRepositoryMetadataResponse,
    ICrudRepositoryMetadataResponseProperty
}
