// @itexpert-dev/base-universal-table/src/directives/universalTableCell/universalTableCell.directive.ts
import {
    Directive,
    ComponentFactoryResolver,
    ViewContainerRef,
    Input,
    Injector
} from "@angular/core";
import {UniversalTableCtrl} from "../../services/universalTable/universalTableCtrl";
import {baseUniversalTableCellDirectiveConfig} from "./baseUniversalTableCellDirectiveConfig.const";
import {BaseUniversalTableCellDirective} from "./BaseUniversalTableCellDirective";
import {UniversalTableCellPosition} from './UniversalTableCellPosition'

@Directive(baseUniversalTableCellDirectiveConfig)
class UniversalTableCellDirective extends BaseUniversalTableCellDirective{
    @Input() rowIndex: number;
    @Input() cellIndex: number;
    @Input() cellData: any;

    constructor(
        componentResolver: ComponentFactoryResolver,
        viewContainer: ViewContainerRef,
        tableCtrl: UniversalTableCtrl,
        injector: Injector
    ){
        super(
            tableCtrl,
            componentResolver,
            viewContainer,
            injector
        )
    }
}

export {UniversalTableCellDirective, UniversalTableCellPosition}
