// @itexpert-dev/base-universal-table/dist/directives/universalTableView.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var universalTableCtrl_1 = require("../services/universalTable/universalTableCtrl");
var BaseUniversalTableViewDirectiveConfig = {
    selector: 'universal-table-view'
};
exports.BaseUniversalTableViewDirectiveConfig = BaseUniversalTableViewDirectiveConfig;
var UniversalTableViewDirective = (function () {
    function UniversalTableViewDirective(componentResolver, viewContainer, injector) {
        this.componentResolver = componentResolver;
        this.viewContainer = viewContainer;
        this.injector = injector;
    }
    UniversalTableViewDirective.prototype.ngOnInit = function () {
        var componentKey = this.tableCtrl.tableComponents.container;
        var componentFactory = this.tableCtrl.config.componentsStorage.getComponent(componentKey);
        var injector = core_1.ReflectiveInjector.resolveAndCreate([{
                provide: universalTableCtrl_1.UniversalTableCtrl,
                useValue: this.tableCtrl
            }], this.injector);
        var component = this.componentResolver.resolveComponentFactory(componentFactory);
        this.viewContainer.createComponent(component, 0, injector);
    };
    return UniversalTableViewDirective;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", universalTableCtrl_1.UniversalTableCtrl)
], UniversalTableViewDirective.prototype, "tableCtrl", void 0);
UniversalTableViewDirective = __decorate([
    core_1.Directive(BaseUniversalTableViewDirectiveConfig),
    __metadata("design:paramtypes", [core_1.ComponentFactoryResolver,
        core_1.ViewContainerRef,
        core_1.Injector])
], UniversalTableViewDirective);
exports.UniversalTableViewDirective = UniversalTableViewDirective;
